### PR TITLE
Jstzd: add rollup build script and spawn rollup in jstzd

### PR DESCRIPTION
**Pull request by ryutamago**
# Context

[spawn rollup in jstzd](https://linear.app/tezos/issue/JSTZ-164/spawn-octezrollupnode-in-jstzd)

NOTE:
The jstz kernel rollup seems to not process the block in the CI (i suppose some issue with `nix.flake`)
the test works fine with the default nix settings `nix flake check`.
I also replaced the jstz kernel with the toy rollup kernel and that worked with the `nix.flake` 

# Description

There are two parts in this PR:
1. add a build script to build the rollup preimages / installer
a. Modified the MakeFile so that the kernel.wasm is copied to crates/jstzd/
b. Add build.rs that 1) stores the rollup kernel_installer.hex, preimages/ and parameters_ty.json to OUT_DIR. 2) generate a path getter file (jstz_rollup_path.rs) to access the path of 1
3. add rollup to the jstzd server

# Manually testing the PR
added rollup to the jstzd integration test.
```
make build
cargo test --package jstzd --test jstzd_test
```
```
nix flake check  
```

### CODE REVIEW

File: crates/jstzd/build_config.rs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+pub const TICKETER_PATH: &[u8] = b"/ticketer";
```
Comment (ryutamago) [inline crates/jstzd/build_config.rs:1]: think we should reexport teh`TICKETER` in jstz_kernel?

File: Makefile [hunk 2]
```diff
@@ -11,13 +11,14 @@ else
 	PROFILE_TARGET_DIR := $(PROFILE)
 endif
 
+JSTZD_KERNEL_PATH := crates/jstzd/resources/jstz_rollup/jstz_kernel.wasm
```
Comment (zcabter) [inline Makefile:14]: Nit: Could we write the kernel to the top level?

File: crates/jstzd/build.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+/// Build script that generates and saves the following files in OUT_DIR:
+///
+/// Files copied:
+/// - parameters_ty.json: JSON file containing parameter types
+///
+/// Files generated:
+/// - kernel_installer.hex: Hex-encoded kernel installer binary
+/// - preimages/: Directory containing kernel preimages
+/// - jstz_rollup_path.rs: Generated Rust code with path getters
```
Comment (zcabter) [inline crates/jstzd/build.rs:33]: 👍

File: crates/jstzd/src/task/jstzd.rs [hunk 4]
```diff
@@ -148,6 +178,18 @@ impl Jstzd {
         }
         Ok(())
     }
+
+    /// Wait for the baker to bake at least `level` blocks.
+    async fn wait_for_block_level(node_endpoint: &Endpoint, level: i64) -> Result<()> {
+        let ready = retry(10, 1000, || async {
+            Ok(get_block_level(&node_endpoint.to_string()).await? >= level)
```
Comment (zcabter) [inline crates/jstzd/src/task/jstzd.rs:185]: Nit: I think you could just map to `bool` like `.await.map(|l| l >= level)` without wrapping in `Ok()`

File: crates/jstzd/src/task/jstzd.rs [hunk 5]
```diff
@@ -148,6 +178,18 @@ impl Jstzd {
...
+        })
+        .await;
+        if !ready {
+            return Err(anyhow!("baker is not ready after retries"));
```
Comment (zcabter) [inline crates/jstzd/src/task/jstzd.rs:189]:
```suggestion
            bail!("baker is not ready after retries");
```

File: crates/jstzd/src/task/jstzd.rs [hunk 6]
```diff
@@ -148,6 +178,18 @@ impl Jstzd {
         }
         Ok(())
     }
+
+    /// Wait for the baker to bake at least `level` blocks.
+    async fn wait_for_block_level(node_endpoint: &Endpoint, level: i64) -> Result<()> {
+        let ready = retry(10, 1000, || async {
```
Comment (zcabter) [inline crates/jstzd/src/task/jstzd.rs:184]: Note: Would be nice if retry returned a `Result<()>` instead. No action required in this PR.

File: crates/jstzd/src/task/utils.rs [hunk 7]
```diff
@@ -13,3 +16,16 @@ where
...
+    let blocks_head_endpoint = format!("{}/chains/main/blocks/head", rpc_endpoint);
+    let response: Value = reqwest::get(&blocks_head_endpoint).await?.json().await?;
+
+    if let Some(level) = response
+        .get("header")
+        .and_then(|header| header.get("level"))
+    {
+        return level.as_i64().ok_or(anyhow!("Level is not a valid i64"));
+    }
+    bail!("Failed to extract level from the JSON response")
```
Comment (zcabter) [inline crates/jstzd/src/task/utils.rs:30]:
```suggestion
    let level = response
        .get("header")
        .and_then(|header| header.get("level"))
        .ok_or_else(|| anyhow!("Failed to extract level from head block"))?;
    level.as_i64().ok_or(anyhow!("Level is not a valid i64"))
```

File: insatller
Comment (zcabter) [inline insatller:1]: File name typo

Comment (zcabter) [reply insatller:1]: Is this still required?

File: Makefile [hunk 8]
```diff
@@ -33,6 +34,11 @@ build-bridge:
 build-kernel:
 	@cargo build --package jstz_kernel --target wasm32-unknown-unknown $(PROFILE_OPT)
 
+.PHONY: build-jstzd-kernel
+build-jstzd-kernel: build-kernel
+	@cp target/wasm32-unknown-unknown/$(PROFILE_TARGET_DIR)/jstz_kernel.wasm $(JSTZD_KERNEL_PATH)
+
+# TODO: Remove once jstzd replaces the sandbox
```
Comment (zcabter) [inline Makefile:41]: Could you track the `TODO` with a linear ticket please?

File: crates/jstzd/tests/jstzd_test.rs [hunk 9]
```diff
@@ -117,14 +116,113 @@ async fn jstzd_test() {
...
+        .set_network("sandbox")
+        .set_rpc_endpoint(octez_node_rpc_endpoint)
+        .set_run_options(&run_options)
+        .build()
+        .unwrap();
+
+    let (rollup_kernel_installer, rollup_preimages_dir, rollup_parameters_ty) =
+        jstz_rollup_files();
+
+    let protocol_params: ProtocolParameter =
```
Comment (zcabter) [inline crates/jstzd/tests/jstzd_test.rs:146]: Rust should be able to infer the type.

File: crates/jstzd/tests/jstzd_test.rs [hunk 10]
```diff
@@ -117,14 +116,113 @@ async fn jstzd_test() {
...
     assert!(!jstzd.baker_healthy().await);
+    let rollup_health = rollup_health_check(&rollup_rpc_endpoint.clone()).await;
+    assert!(rollup_health.is_err());
     assert!(!jstzd.health_check().await);
 
     // stop should be idempotent and thus should be okay after jstzd is already stopped
     jstzd.stop().await.unwrap();
 }
 
+async fn build_configs(
```
Comment (zcabter) [inline crates/jstzd/tests/jstzd_test.rs:128]: Note: These are probably the default values we want to instantiate jstzd with. I expect that when we put the cli together, these values would be moved into `src/` instead.

Comment (huancheng-trili) [reply crates/jstzd/tests/jstzd_test.rs:128]: Yes. I've implemented that actually just waiting for the prerequisite branches to be merged. The default values will be set when the program parses and processes user config.

File: crates/jstzd/tests/jstzd_test.rs [hunk 11]
```diff
@@ -1,104 +1,102 @@
...
+    // check bootstrap contracts
     let octez_client = OctezClient::new(octez_client_config);
     check_bootstrap_contracts(&octez_client).await;
 
+    // check if rollup is running and logging to the debug file
+    let rollup_running = retry(30, 1000, || async {
+        let res = rollup_health_check(&rollup_rpc_endpoint.clone()).await;
+        Ok(res.is_ok_and(|healthy| healthy))
+    })
+    .await;
```
Comment (zcabter) [inline crates/jstzd/tests/jstzd_test.rs:84]:
```suggestion
    let rollup_running =
        retry(30, 1000, || rollup_health_check(&rollup_rpc_endpoint)).await;
```

**Comment by zcabter** [comment]
Btw, you should add `./resources/jstz_rollup` and `jstz_rollup_path.rs` to the `.gitignore` and `include` these files as part of the jstzd package in `Cargo.toml`

File: crates/jstzd/build.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+    // 3. Save hex-encoded kernel installer to OUT_DIR
+    fs::write(out_dir.join("kernel_installer.hex"), kernel_installer)
+        .expect("Failed to write kernel_installer.hex");
+
+    // 4. Generate path getter code in OUT_DIR
+    generate_code(&out_dir);
+
+    println!(
+        "cargo:warning=Build script output directory: {}",
```
[the remark this answers is not in the document]
Comment (zcabter) [reply crates/jstzd/build.rs:64]: My bad, there is no such thing as `cargo:info`

File: Makefile (hunk 8 above)
Comment (ryutamago) [reply Makefile:41]: done

File: crates/jstzd/src/task/jstzd.rs [hunk 13]
```diff
@@ -122,6 +142,16 @@ impl Jstzd {
...
     }
 
+    async fn import_rollup_operator(octez_client: &OctezClient) {
+        octez_client
+            .import_secret_key(
+                Self::ROLLUP_OPERATOR_ACCOUNT_ALIAS,
+                Self::ROLLUP_OPERATOR_ACCOUNT_SK,
+            )
+            .await
+            .expect("Failed to import account 'rollup_operator'");
```
[the remark this answers is not in the document]
Comment (ryutamago) [reply crates/jstzd/src/task/jstzd.rs:152]: good one!

File: crates/jstzd/tests/jstzd_test.rs (hunk 9 above)
Comment (ryutamago) [reply crates/jstzd/tests/jstzd_test.rs:146]: fixed

File: crates/jstzd/src/task/jstzd.rs (hunk 4 above)
Comment (ryutamago) [reply crates/jstzd/src/task/jstzd.rs:185]: neat!

File: crates/jstzd/tests/jstzd_test.rs [hunk 14]
```diff
@@ -1,32 +1,57 @@
...
+    SmartRollupPvmKind,
 };
+use octez::r#async::rollup::{OctezRollupConfigBuilder, RollupDataDir};
 use octez::unused_port;
+use serde_json::from_str;
+use std::fs;
+use tempfile::NamedTempFile;
+use tezos_crypto_rs::hash::SmartRollupHash;
 
+include!(concat!(env!("OUT_DIR"), "/jstz_rollup_path.rs"));
```
Comment (zcabter) [inline crates/jstzd/tests/jstzd_test.rs:25]: Its best to add this line in `src/lib.rs` then import it from there

Comment (ryutamago) [reply crates/jstzd/tests/jstzd_test.rs]: done

File: file_path
Comment (zcabter) [inline file_path:1]: Probably not needed?

File: crates/jstzd/src/config.rs [hunk 15]
```diff
@@ -65,6 +68,21 @@ pub(crate) async fn build_config(
         &octez_client_config,
     )?;
 
+    // dummy rollup config for now
+    // TOOD: https://linear.app/tezos/issue/JSTZ-238/deserialize-rollup-config
```
Comment (zcabter) [inline crates/jstzd/src/config.rs:72]:
```suggestion
    // TODO: https://linear.app/tezos/issue/JSTZ-238/deserialize-rollup-config
    // Dummy rollup config for now
```

File: crates/jstzd/src/task/jstzd.rs [hunk 16]
```diff
@@ -139,12 +157,25 @@ impl Jstzd {
...
+        "unencrypted:edsk3gUfUPyBSfrS9CCgmCiQsTCHGkviBDusMxDJstFtojtc1zcpsh";
+    const ROLLUP_OPERATOR_ACCOUNT_ALIAS: &'static str = "bootstrap1";
 
-    async fn import_activator(octez_client: &OctezClient) {
+    async fn import_activator(octez_client: &OctezClient) -> Result<()> {
         octez_client
             .import_secret_key(Self::ACTIVATOR_ACCOUNT_ALIAS, Self::ACTIVATOR_ACCOUNT_SK)
             .await
-            .expect("Failed to import account 'activator'");
+            .context("Failed to import account 'activator'")
```
Comment (zcabter) [inline crates/jstzd/src/task/jstzd.rs:168]:
```suggestion
            .context(format!("Failed to import account '{}'", Self::ACTIVATOR_ACCOUNT_ALIAS))
```

File: crates/jstzd/src/task/jstzd.rs [hunk 17]
```diff
@@ -139,12 +157,25 @@ impl Jstzd {
...
+    }
+
+    async fn import_rollup_operator(octez_client: &OctezClient) -> Result<()> {
+        octez_client
+            .import_secret_key(
+                Self::ROLLUP_OPERATOR_ACCOUNT_ALIAS,
+                Self::ROLLUP_OPERATOR_ACCOUNT_SK,
+            )
+            .await
+            .context("Failed to import account 'rollup_operator'")
```
Comment (zcabter) [inline crates/jstzd/src/task/jstzd.rs:178]:
```suggestion
            .context(format!("Failed to import account '{}'", Self::ROLLUP_OPERATOR_ACCOUNT_ALIAS))
```

File: crates/jstzd/tests/jstzd_test.rs [hunk 18]
```diff
@@ -64,14 +92,28 @@ async fn create_jstzd_server(
...
+            BootstrapAccount::new(ACTIVATOR_PK, 15_000_000_000).unwrap(),
+            BootstrapAccount::new(JSTZ_ROLLUP_OPERATOR_PK, 60_000_000_000).unwrap(),
+        ])
+        .set_bootstrap_smart_rollups([BootstrapSmartRollup::new(
+            JSTZ_ROLLUP_ADDRESS,
+            SmartRollupPvmKind::Wasm,
+            fs::read_to_string(rollup_kernel_installer.as_path())
+                .unwrap()
+                .as_str(),
+            from_str(fs::read_to_string(rollup_parameters_ty).unwrap().as_str())
```
Comment (zcabter) [inline crates/jstzd/tests/jstzd_test.rs:115]:
```suggestion
            Value::from_str(fs::read_to_string(rollup_parameters_ty).unwrap().as_str())
```

Comment (ryutamago) [reply crates/jstzd/tests/jstzd_test.rs]: done

File: crates/jstzd/src/config.rs (hunk 15 above)
Comment (ryutamago) [reply crates/jstzd/src/config.rs]: done

**Review by zcabter — APPROVED** [review]

**Pull request merged**